File: src/parsing/parser.cpp
/*
Orders Of Precedence https://en.cppreference.com/w/cpp/language/operator_precedence.html

  - Assignment
  - Logic operators (and, or, xor, not)
  - Comparison operators (==, !=, >, <, >=, <=)
  - Bitwise shift
  - AdditiveExpr
  - MultiplicitaveExpr
  - Call   - ~~Member~~ // no members as of now `a().b.c()` - Subscript
  - PrimaryExpr // callable, if, while definitions (because if everything is an expression, these need to be evaled first)

TODOs:

  - OOP (far future)
  - Pattern matching (maybe a new type)
  - Maybe objects
  - Maybe FFI
  - some sort of way to interact with RuntimeVal's (maybe methods `"".join(list)` or modules for it `string.join("", list)`)
*/

#include "lexer.hpp"
#include "ast.hpp"
#include "parser.hpp"
#include "../Errors.hpp"
#include "../util.hpp"

bool Parser::not_eof() {
  return tokens[0].type != TokenType::EndOfFile;
}

Token Parser::curr() {
  return tokens[0];
}

Token Parser::look_ahead(int n) {
  return tokens[n];
}

Token Parser::advance() {
  Token value = tokens.front();
  tokens.pop_front();

  return value;
}

Token Parser::expect(TokenType expected_token, std::string error_message) {
  Token value = advance();

  if (value.type != expected_token) {
    raise_error(error_message);
  }

  return value;
}

OperatorType Parser::get_math_operator(std::string operatorLiteral) {
  if (operatorLiteral == "+") {
    return OperatorType::add;
  } else if (operatorLiteral == "-") {
    return OperatorType::substract;
  } else if (operatorLiteral == "*") {
    return OperatorType::multiply;
  } else if (operatorLiteral == "/") {
    return OperatorType::divide;
  } else if (operatorLiteral == "%") {
    return OperatorType::modulo;
  } else {
    raise_error("Unexpected operator, " + operatorLiteral);
  }
}

LogicalOperatorType Parser::get_logical_operator(std::string operatorLiteral) {
  if (operatorLiteral == "and") {
    return LogicalOperatorType::And;
  } else if (operatorLiteral == "or") {
    return LogicalOperatorType::Or;
  } else if (operatorLiteral == "xor") {
    return LogicalOperatorType::Xor;
  } else {
    raise_error("Unexpected logical operator, " + operatorLiteral);
  }
}

ComparisonOperatorType Parser::get_comparison_operator(std::string operatorLiteral) {
  if (operatorLiteral == "==") {
    return ComparisonOperatorType::equal;
  } else if (operatorLiteral == ">") {
    return ComparisonOperatorType::greater;
  } else if (operatorLiteral == ">=") {
    return ComparisonOperatorType::greater_equal;
  } else if (operatorLiteral == "<") {
    return ComparisonOperatorType::less;
  } else if (operatorLiteral == "<=") {
    return ComparisonOperatorType::less_equal;
  } else if (operatorLiteral == "!=") {
    return ComparisonOperatorType::not_equal;
  } else {
    raise_error("Unexpected logical operator, " + operatorLiteral);
  }
}

Program* Parser::parse_ast(std::string& sourceCode) {
  tokens = tokenize(sourceCode);
  Program* program = new Program();
  while (not_eof()) {
    program->body.push_back(parse_expr()); // everything is an expression :clueless:
  }

  return program;
}

Expr* Parser::parse_expr() {
  return parse_assignment_expr();
}

Expr* Parser::parse_assignment_expr() {
  if (curr().type == TokenType::Const) {
    advance(); // eat const keyword
    Expr* iden = parse_primary_expr();

    if (iden->kind != NodeType::Identifier) {
      raise_error("Expected a identifier after a const keyword");
    }
    expect(TokenType::Equals, "an Equal sign is needed after a const declaration"); // remove the equal sign

    VariableDeclaration* var = new VariableDeclaration();
    var->constant = true;
    var->identifier = static_cast<Identifier*>(iden)->value;
    var->value = parse_expr();
    return var;
  } else if (curr().type == TokenType::Local) {
    advance(); // eat local keyword
    Expr* iden = parse_primary_expr();

    if (iden->kind != NodeType::Identifier) {
      raise_error("Expected a identifier after a const keyword");
    }

    Expr* value;

    if (curr().type == TokenType::Equals) {
      advance();
      value = parse_expr();
    } else {
      Identifier* empty = new Identifier();
      empty->value = "empty";

      value = empty;
    }

    AssignmentExpr* var = new AssignmentExpr();
    var->local = true;
    var->identifier = iden;
    var->value = value;
    return var;
  }
  Expr* left = parse_logical_expr();

  if (curr().type == TokenType::Equals) {
    advance(); // eat the curr equal sign
    Expr* right = parse_expr();

    AssignmentExpr* assign = new AssignmentExpr();
    assign->identifier = left;
    assign->value = right;
    return assign;
  }
  return left;
}

Expr* Parser::parse_logical_expr() {
  if (curr().type == TokenType::Not) {
    advance();
    Expr* negatedExpr = parse_logical_expr();

    NegateExpr* neg = new NegateExpr();
    neg->expr = negatedExpr;
    return neg;
  }

  Expr* left = parse_comparison_expr();
  while (curr().type == TokenType::LogicalExpr) {
    LogicalExpr* logicExpr = new LogicalExpr();

    logicExpr->op = get_logical_operator(curr().value);
    advance();
    logicExpr->left = left;
    logicExpr->right = parse_comparison_expr();


    left = logicExpr;
  }
  return left;
}

Expr* Parser::parse_comparison_expr() {
  Expr* left = parse_bitwise_shift_expr();
  while (curr().type == TokenType::ComparisonExpr) {
    ComparisonExpr* compExpr = new ComparisonExpr();

    compExpr->op = get_comparison_operator(curr().value);
    advance();
    compExpr->left = left;
    compExpr->right = parse_bitwise_shift_expr();

    left = compExpr;
  }
  return left;
}

Expr* Parser::parse_bitwise_shift_expr() {
  Expr* left = parse_additive_expr();
  while (curr().type == TokenType::BitwiseShift) {
    BitShiftExpr* bitShiftExpr = new BitShiftExpr();

    if (advance().value == ">>") {
      bitShiftExpr->shiftRight = true;
    } else {
      bitShiftExpr->shiftRight = false;
    }

    bitShiftExpr->left = left;
    bitShiftExpr->right = parse_additive_expr();

    left = bitShiftExpr;
  }
  return left;
}

Expr* Parser::parse_additive_expr() {
  Expr* left = parse_multiplicative_expr();
  while (curr().value == "+" || curr().value == "-") {
    std::string operatorType = advance().value;
    Expr* right = parse_multiplicative_expr();

    BinaryExpr* binary = new BinaryExpr();
    binary->expr_operator = get_math_operator(operatorType);
    binary->left = left;
    binary->right = right;

    left = binary;
  }
  return left;
}

Expr* Parser::parse_multiplicative_expr() {
  Expr* left = parse_call_member_expr();
  while (curr().value == "*" || curr().value == "/" || curr().value == "%") {
    std::string operatorType = advance().value;
    Expr* right = parse_call_member_expr();

    BinaryExpr* binary = new BinaryExpr();
    binary->expr_operator = get_math_operator(operatorType);
    binary->left = left;
    binary->right = right;

    left = binary;
  }
  return left;
}

Expr* Parser::parse_call_member_expr() {
  Expr* left = parse_primary_expr(); // no members as of now
  while (curr().type == TokenType::OpenParen || curr().type == TokenType::OpenBracket || curr().type == TokenType::Dot) {
    switch (curr().type) {
      case TokenType::OpenParen:
        left = parse_call_expr(left);
        break;
      case TokenType::OpenBracket:
        left = parse_subscript_expr(left);
        break;
      case TokenType::Dot:
        left = parse_member_expr(left);
        break;
      default:
        raise_error("wut2");
    }
  }
  return left;
}

Expr* Parser::parse_subscript_expr(Expr* left) {
  if (curr().type != TokenType::OpenBracket) return left;
  advance();

  SubscriptExpr* sub = new SubscriptExpr();
  sub->left = left;
  sub->value = parse_expr();

  expect(TokenType::ClosedBracket, "expected closing braket on subscript end");
  return sub;
}

Expr* Parser::parse_member_expr(Expr* left) {
  advance(); // eat the dot
  Expr* iden = parse_primary_expr();
  if (iden->kind != NodeType::Identifier)
    raise_error("expected an identifier after the dot expr");

  MemberExpr* memberExpr = new MemberExpr();
  memberExpr->left = left;
  memberExpr->identifier = static_cast<Identifier*>(iden)->value;

  return memberExpr;
}

Expr* Parser::parse_call_expr(Expr* caller) {
  CallExpr* callExpr = new CallExpr();
  callExpr->caller = caller;
  callExpr->args = parse_call_args();

  if (curr().type == TokenType::OpenParen) {
    return parse_call_expr(callExpr);
  }
  return callExpr;
}

std::vector<Expr*> Parser::parse_call_args() {
  advance(); // eat the open paren
  std::vector<Expr*> args = {};
  if (curr().type != TokenType::ClosedParen) {
    Expr* arg = parse_expr();
    args.push_back(arg);

    while (curr().type == TokenType::Comma) {
      advance(); // eat the comma
      Expr* arg = parse_expr();
      args.push_back(arg);
    }
  }
  expect(TokenType::ClosedParen, "Expected a closing paren"); // eat the closing paren
  return args;
}

std::vector<Expr*> Parser::parse_list_elements() {
  advance(); // eat the open braket
  std::vector<Expr*> args;
  if (curr().type != TokenType::ClosedBracket) {
    Expr* arg = parse_expr();
    args.push_back(arg);

    while (curr().type == TokenType::Comma) {
      advance(); // eat the comma
      Expr* arg = parse_expr();
      args.push_back(arg);
    }
  }
  expect(TokenType::ClosedBracket, "Expected a closing bracket"); // eat the closing bracket
  return args;
}

Expr* Parser::parse_special_expr() {
  if (curr().type != TokenType::Identifier)
    raise_error("missing identifier after @");

  Identifier* iden = static_cast<Identifier*>(parse_primary_expr());

  SpecialExpr* specialExpr = new SpecialExpr();
  specialExpr->identifier = iden->value;

  if (curr().type != TokenType::OpenParen) { // if i want to add special value or sum
    raise_error("Expected a opening paren after special function identifier " + iden->value);
  }
  advance();

  specialExpr->isFunction = true;

  while (not_eof() && curr().type != TokenType::ClosedParen) {
    if (curr().type != TokenType::String)
      raise_error("param to the special function should be a string");

    specialExpr->args.push_back(static_cast<StringLiteral*>(parse_primary_expr()));
  }

  expect(TokenType::ClosedParen, "Missing closing paren for special function " + iden->value);

  return specialExpr;
}

Expr* Parser::parse_primary_expr() {
  TokenType token = curr().type;

  switch (token) {
    case TokenType::Identifier: {
      Identifier* iden = new Identifier();
      iden->value = advance().value;

      return iden;
    }
    case TokenType::Number: {
      NumberLiteral* numLit = new NumberLiteral();
      numLit->value = std::stod(advance().value);

      return numLit;
    }
    case TokenType::String: {
      StringLiteral* strLit = new StringLiteral();
      strLit->value = advance().value;

      return strLit;
    }
    case TokenType::OpenParen: {
      advance();
      Expr* value = parse_expr();
      expect(TokenType::ClosedParen, "Expected a closing Parenthesis");
      return value;
    }
    case TokenType::Callable: {
      advance(); // eat the keyword
      if (curr().type != TokenType::OpenParen)
        raise_error("Expected open paren after callable keyword");

      std::vector<std::string> params = {};
      std::vector<Expr*> args = parse_call_args();
      for (auto arg : args) {
        if (arg->kind != NodeType::Identifier) // check if every param is an identifier
          raise_error("Only identifiers in callable definition");
        params.push_back(static_cast<Identifier*>(arg)->value);
      }

      FunctionDeclaration* func = new FunctionDeclaration();
      func->parameters = params;

      expect(TokenType::OpenBrace, "Expected an open Brace for callable body declaration");

      while (not_eof() && curr().type != TokenType::ClosedBrace) {
        Stmt* stmt = parse_expr();
        func->body.push_back(stmt);
      }
      expect(TokenType::ClosedBrace, "Expected a closing brace to close the callable body definition");

      return func;
    }
    case TokenType::If: {
      advance();
      IfStatement* IfExpr = new IfStatement();
      IfExpr->check = parse_expr();

      expect(TokenType::OpenBrace, "Expected an open Brace for if body declaration");

      while (not_eof() && curr().type != TokenType::ClosedBrace) {
        Stmt* stmt = parse_expr();
        IfExpr->body.push_back(stmt);
      }
      expect(TokenType::ClosedBrace, "Expected a closing brace to close the if body definition");

      while (curr().type == TokenType::ElseIf) {
        advance();

        Expr* check = parse_expr();
        std::vector<Stmt*> else_if_body;

        expect(TokenType::OpenBrace, "Expected an open Brace for else_if body declaration");

        while (not_eof() && curr().type != TokenType::ClosedBrace) {
          Stmt* stmt = parse_expr();
          else_if_body.push_back(stmt);
        }
        expect(TokenType::ClosedBrace, "Expected a closing brace to close the else_if body definition");

        IfExpr->else_if_chain.push_back({check, else_if_body});
      }

      if (curr().type == TokenType::Else) {
        advance();
        expect(TokenType::OpenBrace, "Expected an open Brace for else body declaration");

        while (not_eof() && curr().type != TokenType::ClosedBrace) {
          Stmt* stmt = parse_expr();
          IfExpr->else_body.push_back(stmt);
        }
        expect(TokenType::ClosedBrace, "Expected a closing brace to close the else body definition");
      }

      return IfExpr;
    }
    case TokenType::While: {
      advance();
      Expr* check = parse_expr();
      WhileStatement* WhileExpr = new WhileStatement();

      WhileExpr->check = check;

      expect(TokenType::OpenBrace, "Expected an open Brace for if body declaration");

      while (not_eof() && curr().type != TokenType::ClosedBrace) {
        Stmt* stmt = parse_expr();
        WhileExpr->body.push_back(stmt);
      }
      expect(TokenType::ClosedBrace, "Expected a closing brace to close the if body definition");

      return WhileExpr;
    }
    case TokenType::OpenBracket: {
      ArrayLiteral* array = new ArrayLiteral();
      array->elements = parse_list_elements();
      return array;
    }
    case TokenType::Monkey: {
      advance();
      return parse_special_expr();
    }
    default:
      print_token_type(token);
      raise_error("Unexpected token: " + curr().value);
  }
}

File: src/parser.rs
/*
Order of Operations

  - Assignment
  - Logic operators (and, or, xor, not)
  - Comparison operators (==, !=, >, <, >=, <=)
  - Bitwise shift
  - AdditiveExpr
  - MultiplicitaveExpr
  - Call   - ~~Member~~ // no members as of now `a().b.c()` - Subscript
  - PrimaryExpr // callable, if, while definitions (because if everything is an expression, these need to be evaled first)
*/

use std::collections::VecDeque;

use crate::ast_types::*;
use crate::errors::*;
use crate::lexer;
use crate::lexer_types::*;

pub(crate) struct Parser {
    tokens: VecDeque<TokenType>,
}

impl Parser {
    pub(crate) fn new(source_code: &str) -> Result<Self, (Errors, String)> {
        let tokens = lexer::tokenize(source_code)?;
        Ok(Self { tokens })
    }

    #[allow(dead_code)]
    pub(crate) fn dump_tokens(self) -> VecDeque<TokenType> {
        self.tokens
    }

    pub(crate) fn parse(&mut self) -> Result<Ast, (Errors, String)> {
        let mut program_body: Vec<Box<Ast>> = Vec::new();
        while self.tokens.len() > 1 {
            // last one is EOF
            let result = Parser::parse_expression(self)?;
            program_body.push(Box::new(result));
        }
        return Ok(Ast::Program(program_body));
    }

    fn parse_expression(&mut self) -> Result<Ast, (Errors, String)> {
        return Parser::parse_assignment_expr(self);
    }

    fn parse_assignment_expr(&mut self) -> Result<Ast, (Errors, String)> {
        if self.tokens[0] == TokenType::Const {
            self.tokens.pop_front();

            let identifier: Box<str>;
            let result = Parser::parse_primitive_expr(self)?;
            match result {
                Ast::Identifier(iden) => {
                    identifier = iden;
                }
                _ => {
                    return Err((
                        Errors::SyntaxError,
                        "Expected a Identifier after the const keyword".to_owned(),
                    ));
                }
            }
            let right: Ast;
            if self.tokens[0] == TokenType::Equals {
                self.tokens.pop_front(); // pop the equals off
                right = Parser::parse_expression(self)?;
            } else {
                return Err((
                    Errors::SyntaxError,
                    "Constant declaration needs a value to be declared".to_owned(),
                ));
            }
            return Ok(Ast::VariableDeclaration {
                identifier: identifier,
                right: Box::new(right),
                constant: true,
            });
        } else if self.tokens[0] == TokenType::Local {
            self.tokens.pop_front();

            let identifier: Ast;
            let token = Parser::parse_primitive_expr(self)?;
            match token {
                Ast::Identifier(_) => {
                    identifier = token;
                }
                _ => {
                    return Err((
                        Errors::SyntaxError,
                        "Expected a Identifier after the local keyword".to_owned(),
                    ));
                }
            }
            let right: Ast;
            if self.tokens[0] == TokenType::Equals {
                self.tokens.pop_front(); // pop the equals off
                right = Parser::parse_expression(self)?;
            } else {
                right = Ast::Identifier("empty".to_string().into_boxed_str());
            }
            return Ok(Ast::Assignment {
                identifier: Box::new(identifier),
                right: Box::new(right),
                local: true,
            });
        }
        let mut left = Parser::parse_logical_expr(self)?;

        while let Some(token) = self.tokens.front() {
            match token {
                TokenType::Equals => {
                    self.tokens.pop_front();

                    if self.tokens[0] == TokenType::EOF {
                        return Err((
                            Errors::SyntaxError,
                            "missing value after the Comparison operator".to_owned(),
                        ));
                    }

                    let right = Parser::parse_binary_shift(self)?;

                    left = Ast::Assignment {
                        identifier: Box::new(left),
                        right: Box::new(right),
                        local: false,
                    };
                }
                _ => break,
            }
        }
        Ok(left)
    }

    fn parse_logical_expr(&mut self) -> Result<Ast, (Errors, String)> {
        if self.tokens[0] == TokenType::Not {
            self.tokens.pop_front();

            if self.tokens[0] == TokenType::EOF {
                return Err((
                    Errors::SyntaxError,
                    "missing value after the not operator".to_owned(),
                ));
            }

            let right = Parser::parse_comparison_expr(self)?;

            return Ok(Ast::Not(Box::new(right)));
        }

        let mut left = Parser::parse_comparison_expr(self)?;

        while let Some(token) = self.tokens.front() {
            match token {
                TokenType::LogicalExpr(operator) => {
                    let op = *operator;
                    self.tokens.pop_front();

                    if self.tokens[0] == TokenType::EOF {
                        return Err((
                            Errors::SyntaxError,
                            "missing value after the Logical operator".to_owned(),
                        ));
                    }

                    let right = Parser::parse_comparison_expr(self)?;

                    left = Ast::LogicalExpr {
                        operator: op,
                        left: Box::new(left),
                        right: Box::new(right),
                    };
                }
                _ => break,
            }
        }
        Ok(left)
    }

    fn parse_comparison_expr(&mut self) -> Result<Ast, (Errors, String)> {
        let mut left = Parser::parse_binary_shift(self)?;

        while let Some(token) = self.tokens.front() {
            match token {
                TokenType::ComparisonExpr(operator) => {
                    let op = *operator;
                    self.tokens.pop_front();

                    if self.tokens[0] == TokenType::EOF {
                        return Err((
                            Errors::SyntaxError,
                            "missing value after the Comparison operator".to_owned(),
                        ));
                    }

                    let right = Parser::parse_binary_shift(self)?;

                    left = Ast::Comparison {
                        operator: op,
                        left: Box::new(left),
                        right: Box::new(right),
                    };
                }
                _ => break,
            }
        }
        Ok(left)
    }

    fn parse_binary_shift(&mut self) -> Result<Ast, (Errors, String)> {
        let mut left = Parser::parse_aditive_expr(self)?;

        while let Some(token) = self.tokens.front() {
            match token {
                TokenType::BitwiseShift(direction) => {
                    let dir = *direction;
                    self.tokens.pop_front();

                    if self.tokens[0] == TokenType::EOF {
                        return Err((
                            Errors::SyntaxError,
                            "missing value after the BitShift operator".to_owned(),
                        ));
                    }

                    let right = Parser::parse_aditive_expr(self)?;

                    left = Ast::BitwiseShift {
                        shift_direction: dir,
                        left: Box::new(left),
                        right: Box::new(right),
                    };
                }
                _ => break,
            }
        }
        Ok(left)
    }

    fn parse_aditive_expr(&mut self) -> Result<Ast, (Errors, String)> {
        let mut left = Parser::parse_multiplicative_expr(self)?;

        while let Some(token) = self.tokens.front() {
            match token {
                TokenType::BinaryOperator(
                    operator @ (BinaryOperator::Add | BinaryOperator::Substract),
                ) => {
                    let op = *operator;
                    self.tokens.pop_front();

                    if self.tokens[0] == TokenType::EOF {
                        return Err((
                            Errors::SyntaxError,
                            "missing value after the Binary operator".to_owned(),
                        ));
                    }

                    let right = Parser::parse_multiplicative_expr(self)?;

                    left = Ast::BinaryOperator {
                        operator: op,
                        left: Box::new(left),
                        right: Box::new(right),
                    };
                }
                _ => break,
            }
        }
        Ok(left)
    }

    fn parse_multiplicative_expr(&mut self) -> Result<Ast, (Errors, String)> {
        let mut left = Parser::parse_call_member_expr(self)?;

        while let Some(token) = self.tokens.front() {
            match token {
                TokenType::BinaryOperator(
                    operator @ (BinaryOperator::Multiply
                    | BinaryOperator::Divide
                    | BinaryOperator::Modulo),
                ) => {
                    let op = *operator;
                    self.tokens.pop_front();

                    if self.tokens[0] == TokenType::EOF {
                        return Err((
                            Errors::SyntaxError,
                            "missing value after the Binary operator".to_owned(),
                        ));
                    }

                    let right = Parser::parse_call_member_expr(self)?;

                    left = Ast::BinaryOperator {
                        operator: op,
                        left: Box::new(left),
                        right: Box::new(right),
                    };
                }
                _ => break,
            }
        }
        Ok(left)
    }

    fn parse_call_member_expr(&mut self) -> Result<Ast, (Errors, String)> {
        let mut left = Parser::parse_primitive_expr(self)?;
        while self.tokens[0] == TokenType::Dot
            || self.tokens[0] == TokenType::OpenParen
            || self.tokens[0] == TokenType::OpenBracket
        {
            match self.tokens[0] {
                TokenType::Dot => {
                    self.tokens.pop_front(); // pop off the dot

                    let right = Parser::parse_primitive_expr(self)?;

                    if !matches!(right, Ast::Identifier(_)) {
                        return Err((
                            Errors::SyntaxError,
                            "Expected a identifier after the dot expression".to_owned(),
                        ));
                    }

                    left = Ast::MemberExpression {
                        left: Box::new(left),
                        right: Box::new(right),
                    }
                }
                TokenType::OpenBracket => {
                    self.tokens.pop_front(); // pop off the opening bracket

                    if self.tokens[0] == TokenType::ClosedBracket {
                        return Err((
                            Errors::SyntaxError,
                            "Expected an expression inside of the square brackets".to_owned(),
                        ));
                    }

                    let right = Parser::parse_expression(self)?;

                    if self.tokens[0] != TokenType::ClosedBracket {
                        return Err((
                            Errors::SyntaxError,
                            "Expected a closing bracket after the membership expression".to_owned(),
                        ));
                    }
                    self.tokens.pop_front();

                    left = Ast::MemberExpression {
                        left: Box::new(left),
                        right: Box::new(right),
                    }
                }
                TokenType::OpenParen => {
                    self.tokens.pop_front(); // pop off the opening paren

                    if self.tokens[0] == TokenType::ClosedParen {
                        // handle no arguments
                        self.tokens.pop_front(); // pop off the closing paren
                        left = Ast::CallExpression {
                            left: Box::new(left),
                            arguments: Vec::new(),
                        };
                        continue;
                    }

                    let first_arg = Parser::parse_expression(self)?;
                    let mut arg_vec = vec![first_arg];
                    while self.tokens[0] == TokenType::Comma {
                        self.tokens.pop_front();

                        let arg = Parser::parse_expression(self)?;

                        arg_vec.push(arg);
                    }

                    if self.tokens[0] != TokenType::ClosedParen {
                        return Err((
                            Errors::SyntaxError,
                            "Expected a closing paren after the call expression".to_owned(),
                        ));
                    }
                    self.tokens.pop_front(); // pop off the closing paren

                    left = Ast::CallExpression {
                        left: Box::new(left),
                        arguments: arg_vec,
                    }
                }
                _ => {
                    return Err((
                        Errors::SyntaxError,
                        format!("Unexpected token {:?}", self.tokens[0]),
                    ));
                }
            }
        }
        Ok(left)
    }

    fn parse_call_arg(
        &mut self,
        parameters: &mut Vec<(Ast, Option<Ast>)>,
    ) -> Option<(Errors, String)> {
        let param: Ast;
        match Parser::parse_primitive_expr(self) {
            Ok(x) => param = x,
            Err(err) => return Some(err),
        }
        match param {
            Ast::Identifier(_) => {
                if self.tokens[0] != TokenType::Equals {
                    parameters.push((param, None));
                    return None;
                }
                self.tokens.pop_front(); // remove the equals
                match Parser::parse_expression(self) {
                    Ok(x) => parameters.push((param, Some(x))),
                    Err(err) => return Some(err),
                }
            }
            _ => {
                return Some((
                    Errors::SyntaxError,
                    "Expected a identifier in a callable parameter".to_owned(),
                ));
            }
        }
        return None;
    }

    fn parse_primitive_expr(&mut self) -> Result<Ast, (Errors, String)> {
        if let Some(token) = self.tokens.pop_front() {
            match token {
                TokenType::String(string) => Ok(Ast::String(string.clone())),
                TokenType::Int(int) => Ok(Ast::Int(int)),
                TokenType::Float(float) => Ok(Ast::Float(float)),
                TokenType::Identifier(iden) => Ok(Ast::Identifier(iden.clone())),
                TokenType::OpenParen => {
                    let result = Parser::parse_expression(self);
                    if result.is_err() {
                        return result;
                    }

                    if self.tokens[0] != TokenType::ClosedParen {
                        return Err((Errors::SyntaxError, "Expected closing paren".to_owned()));
                    }
                    self.tokens.pop_front();

                    result
                }
                TokenType::If => {
                    let check = Parser::parse_expression(self)?;

                    if self.tokens[0] != TokenType::OpenBrace {
                        return Err((
                            Errors::SyntaxError,
                            "Expected a body after the if check".to_owned(),
                        ));
                    }
                    self.tokens.pop_front(); // pop the opening brace

                    let mut body_vec: Vec<Ast> = Vec::new();
                    while self.tokens[0] != TokenType::ClosedBrace
                        && self.tokens[0] != TokenType::EOF
                    {
                        let expr = Parser::parse_expression(self)?;
                        body_vec.push(expr);
                    }

                    if self.tokens[0] != TokenType::ClosedBrace {
                        return Err((
                            Errors::SyntaxError,
                            "Expected a closing brace for the if block".to_owned(),
                        ));
                    }
                    self.tokens.pop_front();

                    return Ok(Ast::IfExpression {
                        check: Box::new(check),
                        body: body_vec,
                    });
                }
                TokenType::While => {
                    let check = Parser::parse_expression(self)?;

                    if self.tokens[0] != TokenType::OpenBrace {
                        return Err((
                            Errors::SyntaxError,
                            "Expected a body after the while check".to_owned(),
                        ));
                    }
                    self.tokens.pop_front(); // pop the opening brace

                    let mut body_vec: Vec<Ast> = Vec::new();
                    while self.tokens[0] != TokenType::ClosedBrace
                        && self.tokens[0] != TokenType::EOF
                    {
                        let expr = Parser::parse_expression(self)?;
                        body_vec.push(expr);
                    }

                    if self.tokens[0] != TokenType::ClosedBrace {
                        return Err((
                            Errors::SyntaxError,
                            "Expected a closing brace for the while block".to_owned(),
                        ));
                    }
                    self.tokens.pop_front();

                    return Ok(Ast::WhileExpression {
                        check: Box::new(check),
                        body: body_vec,
                    });
                }
                TokenType::Callable => {
                    if self.tokens[0] != TokenType::OpenParen {
                        return Err((
                            Errors::SyntaxError,
                            "Expected Opening Paren after the callable keyword".to_owned(),
                        ));
                    }
                    self.tokens.pop_front(); // eat the open paren

                    let mut parameters: Vec<(Ast, Option<Ast>)> = Vec::new();

                    if self.tokens[0] != TokenType::ClosedParen {
                        if self.tokens[0] == TokenType::EOF {
                            return Err((
                                Errors::SyntaxError,
                                "Expected a closing paren to the callable expression".to_owned(),
                            ));
                        }
                        match Parser::parse_call_arg(self, &mut parameters) {
                            Some(err) => return Err(err),
                            None => {}
                        }
                        while self.tokens[0] != TokenType::ClosedParen
                            && self.tokens[0] == TokenType::Comma
                        {
                            self.tokens.pop_front(); // pop off the comma
                            match Parser::parse_call_arg(self, &mut parameters) {
                                Some(err) => return Err(err),
                                None => {}
                            }
                        }
                    }
                    if self.tokens[0] != TokenType::ClosedParen {
                        return Err((
                            Errors::SyntaxError,
                            "Expected a closing paren to the callable expression".to_owned(),
                        ));
                    }

                    self.tokens.pop_front(); // eat the closed paren

                    if self.tokens[0] != TokenType::OpenBrace {
                        return Err((
                            Errors::SyntaxError,
                            "Expected a body after the callable params".to_owned(),
                        ));
                    }
                    self.tokens.pop_front(); // pop the opening brace

                    let mut body = Vec::new();
                    while self.tokens[0] != TokenType::ClosedBrace
                        && self.tokens[0] != TokenType::EOF
                    {
                        let expr = Parser::parse_expression(self)?;
                        body.push(expr);
                    }

                    if self.tokens[0] != TokenType::ClosedBrace {
                        return Err((
                            Errors::SyntaxError,
                            "Expected a closing brace after the callable block".to_owned(),
                        ));
                    }
                    self.tokens.pop_front();

                    return Ok(Ast::CallableExpression { parameters, body });
                }
                _ => Err((Errors::SyntaxError, format!("Unknown Token {:?}", token))),
            }
        } else {
            Err((Errors::SyntaxError, "Unexpected end of tokens".to_owned()))
        }
    }
}
